### PR TITLE
React capstone project - Metrics webapp

### PR DESCRIPTION
**Pull request by Abdulwhbhussain**
Interactions

Home page

When the page loads, the webapp shows the list of categories that could be filtered out by some parameter, for example by the category name.
Along with the category name, you will display some numeric values per category that come from the API.
When the user clicks (or taps) on a category item, the application navigates to the details page.
Details page

In the details page, the webapp retrieves data from the API to show detailed data in the category.
When the user clicks on the "Back" button (<), the user navigates to the home page.
Testing requirements

Create unit tests for pure functions in your Redux code.
Create integration tests for your application using the React Testing Library.
You may need to mock the access to the API, so that your tests don't send actual requests.
You may need to mock the connection to the Redux Store.
Technical requirements

The project is a single-page application (SPA) built with React and Redux.
The data retrieved from the API should be stored in the Redux store.
You should filter the data that you retrieve from the API using a Filter stateless component.
Every page (the main page and the pages for each item) should have a unique route within the SPA.
The project should be deployed and accessible online using [Render](https://render.com/), [Netlify](https://www.netlify.com/) or [GitHub Pages](https://pages.github.com/).
Add a link to the online version of your app to your README file.
Project documentation
Once you have finished the development of the project, you should record a video presenting the features of the project you built. It is a video with a maximum length of 5 minutes. The content of the video should include:

a description of the project.
a demo of the project features.
you should also highlight some interesting piece of code or something you built that you are very proud of.
For recording the video you can use tools like [Loom](https://www.loom.com/) that let you share a private link to the recording, and configure a shot that shows your computer screen and your face at the same time in a small picture.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,222 @@
...
+
+
+
+<!-- GETTING STARTED -->
+
+## 💻 Getting Started <a name="getting-started"></a>
+
+
+
+> To get a local copy up and running, follow these steps.
```
Comment (Shedrack-Sunday) [inline README.md:116]: - [x] Based on project requirements, you're expected to create a video with the requirements below. Kindly review and do so 👍 

### In
image
structions:

File: README.md [hunk 2]
```diff
@@ -1,2 +1,222 @@
...
+
+### Run tests
+
+> npm run test
+
+### Deployment
+
+> Here is the Link to this Live Project:
+
+> https://space-travelers-hub-three.vercel.app/
```
Comment (Shedrack-Sunday) [inline README.md]: - [x] This link is for `Space Travellers` project and not the metrics app. Kindly deploy and add the appropriate link 👍

File: README.md [hunk 3]
```diff
@@ -1,2 +1,222 @@
...
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- ACKNOWLEDGEMENTS -->
+
+## 🙏 Acknowledgments <a name="acknowledgements"></a>
+
+I would like to thank Microverse and The Code Reviewers of This Project for giving 
+Reviews and Feedback Regarding This Project. I would like to thank Microverse for this opportunity. Ko Kaung Myat Kyaw(Berry) for Design ideas.
+
```
Comment (Shedrack-Sunday) [inline README.md]: - [x] The right author is Nelson Sakwa. Kindly edit this section to give him credit 👍 

### Instructions:
image

File: src/App.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  return (
+    <Router>
+      <Navbar />
+      <Routes>
+        <Route path="/" element={<Home />} />
+        <Route path="/details/:name/:lat/:lon" element={<Details />} />
+        <Route path="/search/:name" element={<SearchPage />} />
+        <Route path="*" element={<ErrorPage />} />
+      </Routes>
+    </Router>
```
Comment (Shedrack-Sunday) [inline src/App.js:24]: - [ ] Your page does not render a design anywhere near the requirements. Kindly review your components, and `CSS` to make sure you have a view similar to the [Design Guildelines](https://www.behance.net/gallery/31579789/Ballhead-App-(Free-PSDs)) 👍 
- [ ] The next review will look at the whole work 

### View:
image

**Review by Shedrack-Sunday — CHANGES_REQUESTED** [review]
Hello @Abdulwhbhussain          👋 

# CHANGES REQUIRED ♻️

Impressive work on the project so far👏
You've done extremely well so far in this project,  However, there are some adjustments you need to make to move to the next project. 💯



## Highlights   
- Good use of GitHub features ✔️   
- Nice work with the Redux features ✔️ 
    

# Required Changes  🌵  

- [x] Firstly, you need to add tests. Both unit and integration tests are as instructed in the project requirements. Kindly do so  👍 
### Instructions:
image


_Check the comments for a detailed review 

# Optional suggestions 

Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing really improves your application.

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: README.md (hunk 2 above)
Comment (Abdulwhbhussain) [reply README.md]: Ok sure!

File: README.md (hunk 3 above)
Comment (Abdulwhbhussain) [reply README.md]: Good Done!

File: README.md (hunk 1 above)
Comment (Abdulwhbhussain) [reply README.md:116]: Done!

File: src/__test__/SearchPage.test.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+          },
+
+        ],
+      },
+    };
+
+    const store = mockStore(initialState);
+    const tree = renderer.create(
+      <Provider store={store}>
+        <MemoryRouter initialEntries={['/search/City Name']}>
```
Comment (thecodechaser) [inline src/__test__/SearchPage.test.js]: - [ ] Good job adding the tests as requested by the previous reviewer. However, this test is failing. Please make sure to fix it and all tests should pass without any errors.

image

File: README.md [hunk 6]
```diff
@@ -1,2 +1,232 @@
...
+<!-- LIVE DEMO -->
+
+## 🚀 Live Demo <a name="live-demo"></a>
+
+- [Live Demo Link](https://www.loom.com/share/bf34dfe621194fcd8e2c0e9599a321ae?sid=9a9c451b-f469-4900-bb8f-099f419c19cb)
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+
```
Comment (thecodechaser) [inline README.md]: - [ ] As requested by the previous reviewer in the presentation video you should also highlight some interesting piece of code or something you built that you are very proud of. You should re-record the video again with all the below-mentioned steps.

image

File: src/pages/Home.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+          <button type="submit" onClick={handleSubmit} className=" md:hidden block hover:bg-sky-700 text-white bg-sky-800 rounded-md py-1 px-3">Search</button>
+        </Search>
+      </div>
+      <div className="city-container grid grid-cols-2 md:grid-cols-5 border border-t-white">
+        {citiesData.map((city) => <CityCard key={city.name} city={city} />)}
+      </div>
+    </div>
+
+  );
+}
```
Comment (thecodechaser) [inline src/pages/Home.js]: - [ ] As requested by the previous reviewer to follow the design guidelines, Please fix the below design issues.
  - Please make sure to display two items(countries) in a row. Currently, It's displaying only one item.
  - Please make sure to display a small button on each item as on the given design, That will indicate users going to the details page.

### Your design

image


### Given design

image

- [ ] Please fix the below header and make sure it matches the given design.

image

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @Abdulwhbhussain👋, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## To Highlight 🎉
- Highly organized code. ✔️
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/__test__/__snapshots__/Detailes.test.js.snap [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      Object {
+        "color": undefined,
+      }
+    }
+    viewBox="0 0 1024 1024"
+    width="1em"
+    xmlns="http://www.w3.org/2000/svg"
+  >
+    <path
+      d="M512 1024c-69.1 0-136.2-13.5-199.3-40.2C251.7 958 197 921 150 874c-47-47-84-101.7-109.8-162.7C13.5 648.2 0 581.1 0 512c0-19.9 16.1-36 36-36s36 16.1 36 36c0 59.4 11.6 117 34.6 171.3 22.2 52.4 53.9 99.5 94.3 139.9 40.4 40.4 87.5 72.2 139.9 94.3C395 940.4 452.6 952 512 952c59.4 0 117-11.6 171.3-34.6 52.4-22.2 99.5-53.9 139.9-94.3 40.4-40.4 72.2-87.5 94.3-139.9C940.4 629 952 571.4 952 512c0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 0 0-94.3-139.9 437.71 437.71 0 0 0-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.2C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3s-13.5 136.2-40.2 199.3C958 772.3 921 827 874 874c-47 47-101.8 83.9-162.7 109.7-63.1 26.8-130.2 40.3-199.3 40.3z"
```
Comment (ghost) [inline src/__test__/__snapshots__/Detailes.test.js.snap:23]: - [ ] Kindly try as much as you can to follow the design guidelines. You do not have to implement the designs 100% perfectly but your design should align with the design guidelines in terms of Typography ( font face, size, and weight) and Layout(composition and space between elements). This is important for consistency throughout your design. Currently, your texts appear to be overflowing with the cards. Kindly check out that

image

File: src/__test__/Home.test.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+describe('Home component', () => {
+  test('Home component renders correctly', () => {
+    const tree = renderer.create(
+      <MemoryRouter>
+        <Home />
+      </MemoryRouter>,
+    ).toJSON();
+    expect(tree).toMatchSnapshot();
+  });
```
Comment (ghost) [inline src/__test__/Home.test.js:14]: - [ ] Your test cases are failing. Kindly check them out.

image

**Review by frank1738 — CHANGES_REQUESTED** [review]
Hi  @Abdulwhbhussain , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!




## Highlights

✅ Correct use of Github workflow
✅ No linter error


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

N/A
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/__test__/Home.test.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+import Home from '../pages/Home';
+
+describe('Home component', () => {
+  test('Home component renders correctly', () => {
+    const tree = renderer.create(
+      <MemoryRouter>
+        <Home />
+      </MemoryRouter>,
+    ).toJSON();
+    expect(tree).toMatchSnapshot();
```
Comment (DuaneDave) [inline src/__test__/Home.test.js:13]: Hey @Abdulwhbhussain, you have done pretty well writing test cases for this component, however, for some reason one test case is not passing and I recommend reviewing the results to spot the problem and resolve them before requesting another review. Doing so would increase your confidence in your code and it s correctness.

image

File: src/components/CityCard.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+const Card = styled.div`
+  border: 2px solid grey;
+  border-radius: 10px;
+  width: 48%;
+  margin: 10px;
+  figure {
+    img {
+      width: 100%;
```
Comment (DuaneDave) [inline src/components/CityCard.js:86]: Ensure to resolve the overflowing text for the city card on the mobile breakpoint as it is not providing the best UX for your app users. Consider removing the set with 48% on it as this is what is causing the overflow. You can set a width of 100% and a max width of 500px to resolve the issue.

image

**Review by DuaneDave — CHANGES_REQUESTED** [review]
Hi @Abdulwhbhussain 👋, 

Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.


## Suggested changes
_Check the comments under the review._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: src/__test__/Home.test.js (hunk 10 above)
Comment (Abdulwhbhussain) [reply src/__test__/Home.test.js:13]: fixed Bro!

File: src/components/CityCard.js (hunk 11 above)
Comment (Abdulwhbhussain) [reply src/components/CityCard.js:86]: Fixed Bro!

**Review by ridwanediallo — APPROVED** [review]
# STATUS: APPROVED 🟢
## Hi @Abdulwhbhussain , 
Congratulations!GIF

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used GitHub flow correctly. ✔
- Well done implementing the requested changes.  🎉 🎉 🎉 ✔

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**